### PR TITLE
Add Replica Capacity Goal.

### PR DESCRIPTION
**Pull request by efeg**
_ReplicaCapacityGoal_ is a hard goal.
It ensures that the maximum number of replicas per broker is under the specified maximum limit controlled by `max.replicas.per.broker`

### CODE REVIEW

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/ReplicaCapacityGoal.java [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+
+   * @param excludedTopics The topics that should be excluded from the optimization proposals.
+   */
+  @Override
+  protected void rebalanceForBroker(Broker broker, ClusterModel clusterModel, Set<Goal> optimizedGoals,
+      Set<String> excludedTopics) throws AnalysisInputException, ModelInputException {
+    LOG.debug("balancing broker {}, optimized goals = {}", broker, optimizedGoals);
+    int numBrokerReplicas = broker.replicas().size();
+    for (Replica replica : new ArrayList<>(broker.replicas())) {
+      if ((broker.isAlive() && numBrokerReplicas <= _balancingConstraint.maxReplicasPerBroker())
```
Comment (becketqin) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/ReplicaCapacityGoal.java]: Should this be checked before iterating over the replicas in the broker?

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/ReplicaCapacityGoal.java [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+          || excludedTopics.contains(replica.topicPartition().topic())) {
+        continue;
+      }
+
+      // The goal requirements are violated. Move replica to an available broker.
+      if (maybeApplyBalancingAction(clusterModel, replica, replicaCapacityEligibleBrokers(replica, clusterModel),
+          BalancingAction.REPLICA_MOVEMENT, optimizedGoals) == null) {
+        throw new AnalysisInputException("Violated replica capacity requirement for broker with id " + broker.id() + ".");
```
Comment (becketqin) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/ReplicaCapacityGoal.java]: Should this be OptimizationFailedException?

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/ReplicaCapacityGoal.java [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+      }
+    }
+    if (!broker.isAlive() && broker.replicas().size() > 0) {
+      throw new AnalysisInputException("THERE ARE REPLICAS ON DEAD BROKER");
```
Comment (becketqin) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/ReplicaCapacityGoal.java]: OptimizationFailedException?

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/ReplicaCapacityGoal.java [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+   * Get a list of replica capacity eligible brokers for the given replica in the given cluster. A broker is eligible
+   * for a given replica if the broker contains less than allowed maximum number of replicas.
+   *
+   * @param replica      Replica for which a set of replica capacity eligible brokers are requested.
+   * @param clusterModel The state of the cluster.
+   * @return A list of replica capacity eligible brokers for the given replica in the given cluster.
+   */
+  private List<Broker> replicaCapacityEligibleBrokers(Replica replica, ClusterModel clusterModel) {
+    // Populate partition rack ids.
+    List<Broker> eligibleBrokers = new ArrayList<>();
```
Comment (becketqin) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/ReplicaCapacityGoal.java]: Could be a TreeSet. Performance wise it should be slightly better.

File: cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/analyzer/DeterministicClusterTest.java [hunk 5]
```diff
@@ -70,8 +72,7 @@ public DeterministicClusterTest(BalancingConstraint balancingConstraint, Cluster
    * @return Parameters for the {@link OptimizationVerifier}.
    */
   @Parameterized.Parameters
-  public static Collection<Object[]> data()
-      throws AnalysisInputException, ModelInputException {
+  public static Collection<Object[]> data() throws AnalysisInputException, ModelInputException, IOException {
```
Comment (becketqin) [inline cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/analyzer/DeterministicClusterTest.java]: Why IOException?

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/ReplicaCapacityGoal.java (hunk 3 above)
Comment (efeg) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/ReplicaCapacityGoal.java]: Redundant check -- removed.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/ReplicaCapacityGoal.java (hunk 1 above)
Comment (efeg) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/ReplicaCapacityGoal.java]: Split the check into the cases for `break` and `continue`. (1) breaks when the alive broker is balanced. (2) continues when the topic is excluded but the first constraint is not satisfied.

File: cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/analyzer/DeterministicClusterTest.java (hunk 5 above)
Comment (efeg) [reply cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/analyzer/DeterministicClusterTest.java]: Redundant -- removed.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/ReplicaCapacityGoal.java [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+        break;
+      }
+      if (excludedTopics.contains(replica.topicPartition().topic())) {
+        continue;
+      }
+
+      // The goal requirements are violated. Move replica to an available broker.
+      if (maybeApplyBalancingAction(clusterModel, replica, replicaCapacityEligibleBrokers(replica, clusterModel),
+          BalancingAction.REPLICA_MOVEMENT, optimizedGoals) == null) {
+        throw new OptimizationFailureException("Violated replica capacity requirement for broker with id " + broker.id() + ".");
```
Comment (becketqin) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/ReplicaCapacityGoal.java]: We may not want to throw exception if the current replica cannot be moved. It is possible that another replica can be moved away if we continue.

**Comment by becketqin** [comment]
Thanks for the patch. LGTM.

**Pull request merged**